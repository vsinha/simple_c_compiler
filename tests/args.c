
void compute(long a, long b)
{
	printf("a=%d b=%d\n", a, b);
}

void main()
{
	compute(6, 7);
}

