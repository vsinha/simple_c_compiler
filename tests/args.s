	.text
.globl compute
.type compute, @function
compute:
	# Save registers
	pushq %rbx
	pushq %r10
	pushq %r13
	pushq %r14
	pushq %r15

	# move the stack pointer for 32 vars of space
	subq $256, %rsp
	# Save arguments
	movq %rdi, 0(%rsp)
	movq %rsi, 8(%rsp)
	#top=0

	# push string "a=%d b=%d\n" top=0
	movq $string0, %rbx
	# 2 top = 1
	# variable a is at 0 position in the list
	movq 0(%rsp), %r10 # load a from local vars
	# 3 top = 2
	# variable b is at 1 position in the list
	movq 8(%rsp), %r13 # load b from local vars
	# 3 top = 3

	# func=printf nargs=3
	# Move values from reg stack to reg args
# 3 top-- = 2
	movq %r13, %rdx
# 3 top-- = 1
	movq %r10, %rsi
# 3 top-- = 0
	movq %rbx, %rdi
	movl $0, %eax # return val for printf
	call printf
	movq %rax, %rbx
	# 1 top = 1
# 18 top-- = 0

	# move the stack pointer back to where it was
	addq $256, %rsp
	# Restore registers
	popq %r15
	popq %r14
	popq %r13
	popq %r10
	popq %rbx
	ret
	.text
.globl main
.type main, @function
main:
	# Save registers
	pushq %rbx
	pushq %r10
	pushq %r13
	pushq %r14
	pushq %r15

	# move the stack pointer for 32 vars of space
	subq $256, %rsp
	# Save arguments

	# push 6 heere
	movq $6,%rbx
	# 7 top = 1

	# push 7 heere
	movq $7,%r10
	# 7 top = 2

	# func=compute nargs=2
	# Move values from reg stack to reg args
# 3 top-- = 1
	movq %r10, %rsi
# 3 top-- = 0
	movq %rbx, %rdi
	call compute
	movq %rax, %rbx
	# 1 top = 1
# 18 top-- = 0

	# move the stack pointer back to where it was
	addq $256, %rsp
	# Restore registers
	popq %r15
	popq %r14
	popq %r13
	popq %r10
	popq %rbx
	ret
string0:
	.string "a=%d b=%d\n"

