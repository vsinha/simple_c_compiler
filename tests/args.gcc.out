args.c: In function 'compute':
args.c:4:2: warning: incompatible implicit declaration of built-in function 'printf' [enabled by default]
args.c:4:2: warning: format '%d' expects argument of type 'int', but argument 2 has type 'long int' [-Wformat]
args.c:4:2: warning: format '%d' expects argument of type 'int', but argument 3 has type 'long int' [-Wformat]
